### PR TITLE
Initial

### PR DESCRIPTION
**Pull request by KlausPopp**
First revision. Base functions complete.

### CODE REVIEW

File: pkg/io4edge/channel.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+type Channel struct {
+	ms transport.MsgStream
+}
+
+// NewChannel creates a new channel using the transport mechanism in t
+func NewChannel(ms transport.MsgStream) (*Channel, error) {
+	return &Channel{ms: ms}, nil
+}
+
+// Close closes the message stream
```
Comment (batthebee) [inline pkg/io4edge/channel.go:37]: remove "Close "

File: pkg/io4edge/fwpkg/fwpkg.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+func NewFirmwarePackageFromFile(fileName string) (*FirmwarePackage, error) {
+	p := new(FirmwarePackage)
+	p.fileName = fileName
+
+	f, err := os.Open(fileName)
+	if err != nil {
+		return nil, errors.New("can't open " + fileName + ": " + err.Error())
+	}
+	p.file = f
+
```
Comment (batthebee) [inline pkg/io4edge/fwpkg/fwpkg.go:56]: loadManifest instead of getManifest ?!

File: pkg/io4edge/fwpkg/tar.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+// and writes the file content to w
+// fileName must be the whole file name with path as in the archive e.g. "./test.txt"
+func untarFileContent(r io.Reader, fileName string, w io.Writer) error {
+
+	tr := tar.NewReader(r)
+
+	for {
+		header, err := tr.Next()
+
+		switch {
```
Comment (batthebee) [inline pkg/io4edge/fwpkg/tar.go:35]: if err {
    if err == io.EOF {
       return new error
   }
  return err
}

Comment (batthebee) [reply pkg/io4edge/fwpkg/tar.go:35]: use if instead of case

File: pkg/io4edge/fwpkg/tar.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+			return err
+
+		case header == nil:
+			continue
+		}
+
+		if header.Typeflag == tar.TypeReg {
+			if header.Name == fileName {
+				if _, err := io.Copy(w, tr); err != nil {
+					return err
```
Comment (batthebee) [inline pkg/io4edge/fwpkg/tar.go:49]: remove if, instead simply return err

File: pkg/io4edge/fwpkg/tar.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+		case err != nil:
+			return err
+
+		case header == nil:
+			continue
+		}
+
+		if header.Typeflag == tar.TypeReg {
+			if header.Name == fileName {
```
Comment (batthebee) [inline pkg/io4edge/fwpkg/tar.go:47]: one line instead of two

File: pkg/io4edge/transport/socket.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return l, nil
+}
+
+// WaitForSocketConnect waits for a client to connect to the TCP socket and returns the TCP connection
+// There is no timeout
+func WaitForSocketConnect(l *net.TCPListener) (*net.TCPConn, error) {
+
```
Comment (batthebee) [inline pkg/io4edge/transport/socket.go:42]: remove line

File: pkg/io4edge/transport/socket.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+	conn, err := net.DialTCP("tcp", nil, addr)
+	if err != nil {
+		return nil, err
+	}
+
+	return conn, nil
+}
+
+// NewMsgStreamFromConnection creates a message stream from TCP connection
+func NewMsgStreamFromConnection(conn *net.TCPConn) (*FramedStream, error) {
```
Comment (batthebee) [inline pkg/io4edge/transport/socket.go:66]: error needed?

**Review by batthebee — COMMENTED** [review]
looks good! see comments and decide by yourself :-)

File: pkg/io4edge/channel.go (hunk 1 above)
Comment (siredmar) [reply pkg/io4edge/channel.go:37]: don't remove "Close ", better rephrase: "// Close terminates the message stream"

**Review by FranziskaKnoedtel — APPROVED** [review]

**Pull request merged**